packages/preferences/src/browser/folder-preference-provider.ts

```diff
@@ -0,0 +1,73 @@
+/********************************************************************************
+ * Copyright (C) 2018 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable } from 'inversify';
+import URI from '@theia/core/lib/common/uri';
+import { PreferenceScope, PreferenceProvider } from '@theia/core/lib/browser';
+import { AbstractResourcePreferenceProvider } from './abstract-resource-preference-provider';
+import { FileSystem, FileStat } from '@theia/filesystem/lib/common';
+
+export const FolderPreferenceProviderFactory = Symbol('FolderPreferenceProviderFactory');
+export interface FolderPreferenceProviderFactory {
+    (options: FolderPreferenceProviderOptions): FolderPreferenceProvider;
+}
+
+export const FolderPreferenceProviderOptions = Symbol('FolderPreferenceProviderOptions');
+export interface FolderPreferenceProviderOptions {
+    folder: FileStat;
+}
+
+@injectable()
+export class FolderPreferenceProvider extends AbstractResourcePreferenceProvider {
+
+    private folderUri: URI | undefined;
+
+    constructor(
+        @inject(FolderPreferenceProviderOptions) protected readonly options: FolderPreferenceProviderOptions,
+        @inject(FileSystem) protected readonly fileSystem: FileSystem
+    ) {
+        super();
+    }
+
+    get uri(): URI | undefined {
+        return this.folderUri;
+    }
+
+    async getUri(): Promise<URI | undefined> {
+        this.folderUri = new URI(this.options.folder.uri);
+        if (await this.fileSystem.exists(this.folderUri.toString())) {
+            const uri = this.folderUri.resolve('.theia').resolve('settings.json');
+            return uri;
+        }
+    }
+
+    canProvide(preferenceName: string, resourceUri?: string): { priority: number, provider: PreferenceProvider } {
+        const value = this.get(preferenceName);
+        if (value === undefined || value === null || !resourceUri || !this.folderUri) {
+            return super.canProvide(preferenceName, resourceUri);
+        }
+        const uri = new URI(resourceUri);
+        return { priority: 3 + this.folderUri.path.relativity(uri.path), provider: this };
```

Comment:
priority value should be defined in a central location, apply to user and workspace as well

Reply:
yes, i will add a "PreferenceProviderPriority" namespace to preference-provider.ts.